add a test for grad view with torch amp

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54995 static graph api skeleton
* #54981 refactor autograd_hook
* **#56730 add a test for grad view with torch amp**
* #56641 change ddpLoggingData struct to map or dict

add a test to verify DDP with torch map will result in the same results when using grad_as_bucket_view=true and false.

torch.amp scale factor does not have dependencies on old gradients, thus it is not affected by grad_as_bucket_view=true or false, see
how torch.amp is implemeted here https://github.com/pytorch/pytorch/pull/33366/files.

This diff verified ddp can work as expected with amp.GradScaler and amp.autocast when when using grad_as_bucket_view=true and false.

Differential Revision: [D27950132](https://our.internmc.facebook.com/intern/diff/D27950132/)

zhaojuanmao committed:
add a test for grad view with torch amp

add a test to verify DDP with torch map will result in the same results when using grad_as_bucket_view=true and false.

torch.amp scale factor does not have dependencies on old gradients, thus it is not affected by grad_as_bucket_view=true or false, see
how torch.amp is implemeted here https://github.com/pytorch/pytorch/pull/33366/files.

This diff verified ddp can work as expected with amp.GradScaler and amp.autocast when when using grad_as_bucket_view=true and false.

Differential Revision: [D27950132](https://our.internmc.facebook.com/intern/diff/D27950132/)

[ghstack-poisoned]

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -290,15 +290,6 @@ class DistributedDataParallel(Module):
-        Mixed Precision (AMP). AMP maintains stashed gradients that are used for
-        unscaling gradients. With ``gradient_as_bucket_view=True``, these
-        stashed gradients will point to communication buckets in the first
-        iteration. In the next iteration, the communication buckets are mutated
```
Could you clarify how this issue was fixed so that the stashed gradients are no longer mutated?

zhaojuanmao replied:
the new amp scaler implementation in PyTorch does not use stashed gradients at all,  so it is not affected by how gradient is calculated

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -3288,6 +3289,57 @@ def test_DistributedDataParallel_with_grad_is_view(self):
+                # scaler.step() first unscales the gradients of the optimizer's assigned params.
+                # If these gradients do not contain infs or NaNs, optimizer.step() is then called,
+                # otherwise, optimizer.step() is skipped.
+                scaler.step(optimizer)
```
Docs state that:
```
If no inf/NaN gradients are found, invokes optimizer.step() using the unscaled gradients. Otherwise, optimizer.step() is skipped to avoid corrupting the params.
```

However, if we skip all optimizer steps in this test, then the DDP bucket view vs not bucket view models would be equivalent since their params weren't modified at all.

Should we add some logic to verify that there are no NaN/inf grads and that parameters are indeed updated after every iteration?

zhaojuanmao replied:
sure, let me add a validation to make sure params are actually changed and updated after training. But be noted that: based on the amp implementation, the first a few iterations could have NaN/inf unscaled gradients, and then gradually the scale factor will be tuned and make the gradients not to be NaN/inf

rohan-varma reviewed (approved):
Thanks for adding these tests! Had 2 questions inline, but overall looks good.

zhaojuanmao committed:
Update on "add a test for grad view with torch amp"


add a test to verify DDP with torch map will result in the same results when using grad_as_bucket_view=true and false.

torch.amp scale factor does not have dependencies on old gradients, thus it is not affected by grad_as_bucket_view=true or false, see
how torch.amp is implemeted here https://github.com/pytorch/pytorch/pull/33366/files.

This diff verified ddp can work as expected with amp.GradScaler and amp.autocast when when using grad_as_bucket_view=true and false.

Differential Revision: [D27950132](https://our.internmc.facebook.com/intern/diff/D27950132/)

[ghstack-poisoned]
